### PR TITLE
Twitter bot

### PR DESCRIPTION
**Pull request by lalalune**
This covers the rest of the functionality for the Twitter bot.

The only main functionality not handled here is trading, since we will need to resolve what that interaction would actually look like. 

Gif previews are also not currently supported.

Everything else should work the same as the Discord bot, although some of the text has been formatted to be a bit friendlier and more informative.

Also, currently "help" links to the discord bot documentation, this will need to be updated.

You will need to add production keys to test, into the config.json.

These features *should* work but need testing by someone with production credentials
play
unsell
buy
store

### CODE REVIEW

**Comment by avaer** [comment]
Do we need `tmp`? Can't this all be done in memory?

**Comment by lalalune** [comment]
I could work on it but piping the output of one request into the input of
another is not something I've ever done, and I was having issues it with so
I figured I'd cache the jpg for a second before uploading and deleting it,
since FS makes sense to me :) but yeah, definitely better to do in memory
if possible.

**Comment by vyrnix9582046173** [comment]
Could it be kept as a file blob (https://developer.mozilla.org/en-US/docs/Web/API/Blob) until it is minted?

**Comment by avaer** [comment]
> Could it be kept as a file blob

No because this is node which does not have official `Blob`.

This code can be copied instead: https://github.com/webaverse/ethereum-bot/blob/27c4782f81735de92c0fd5fe9923dd793ad65008/discordbot.js#L2231

The traditional way is to flow the request/response.

File: twitterBot.js [hunk 1]
```diff
@@ -1,105 +1,1766 @@
...
+    console.log(media_tmp);
+  }
+
+  if (typeof media_tmp !== 'undefined') {
+    const newUrl = media_tmp.media_url ?? media_tmp.media.media_url;
+    const filename = newUrl.split("/")
+    const resourcePath = `http://localhost:${serverPort}/${filename[filename.length - 1]}`
+    const filePath = `tmp/${filename[filename.length - 1]}`
+
+    downloadMedia(newUrl, filePath, function () {
```
Comment (avaer) [inline twitterBot.js]: This could be async.

File: twitterBot.js [hunk 2]
```diff
@@ -1,105 +1,1766 @@
...
+      manualUrl = resourcePath;
+      finishMinting(id, twitterUserId, manualUrl, quantity, event, messageType, filePath)
```
Comment (avaer) [inline twitterBot.js]: `await`?

**Comment by avaer** [comment]
This there a way to do this without `ngrok`?

File: twitterBot.js [hunk 3]
```diff
@@ -1,105 +1,1766 @@
...
+    console.log(media_tmp);
+  }
+
+  if (typeof media_tmp !== 'undefined') {
+    const newUrl = media_tmp.media_url ?? media_tmp.media.media_url;
+    const filename = newUrl.split("/")
+    const resourcePath = `http://localhost:${serverPort}/${filename[filename.length - 1]}`
+    const filePath = `tmp/${filename[filename.length - 1]}`
+
+    downloadMedia(newUrl, filePath, function () {
```
Comment (avaer) [inline twitterBot.js]: Could be `await`.

File: twitterBot.js [hunk 4]
```diff
@@ -1,105 +1,1766 @@
...
+      manualUrl = resourcePath;
+      finishUpdating(id, twitterUserId, manualUrl, nftId, event, messageType, filePath)
```
Comment (avaer) [inline twitterBot.js]: `await`

File: twitterBot.js [hunk 5]
```diff
@@ -1,105 +1,1766 @@
...
+    const filename = newUrl.split("/")
+    const resourcePath = `http://localhost:${serverPort}/${filename[filename.length - 1]}`
+    const filePath = `tmp/${filename[filename.length - 1]}`
+
+    downloadMedia(newUrl, filePath, function () {
+      manualUrl = resourcePath;
+      finishMinting(id, twitterUserId, manualUrl, quantity, event, messageType, filePath)
+    })
+  } else {
+    finishMinting(id, twitterUserId, manualUrl, quantity, event, messageType)
```
Comment (avaer) [inline twitterBot.js]: `await`

File: twitterBot.js [hunk 6]
```diff
@@ -1,105 +1,1766 @@
...
+    const filename = newUrl.split("/")
+    const resourcePath = `http://localhost:${serverPort}/${filename[filename.length - 1]}`
+    const filePath = `tmp/${filename[filename.length - 1]}`
+
+    downloadMedia(newUrl, filePath, function () {
+      manualUrl = resourcePath;
+      finishUpdating(id, twitterUserId, manualUrl, nftId, event, messageType, filePath)
+    })
+  } else {
+    finishUpdating(id, twitterUserId, manualUrl, nftId, event, messageType)
```
Comment (avaer) [inline twitterBot.js]: `await`

**Comment by avaer** [comment]
This overall looks good, just the questions of:

- avoiding disk
- necessity of `ngrok` (less important, would need to set up our own keys for deploy)

File: twitterBot.js [hunk 7]
```diff
@@ -1,105 +1,1766 @@
...
+var static = require('node-static');
+var fileServer = new (static.Server)(__dirname + '/tmp');
+
+const { twitterUsersTableName, usersTableName, storageHost, previewHost, previewExt } = require('./constants')
+
+let ddb, web3, contracts, getStores, runSidechainTransaction = null;
 
 const {
+  treasuryMnemonic,
+  twitterBearerToken,
```
Comment (avaer) [inline twitterBot.js]: `twitterBearerToken` looks unused.

**Review by avaer — CHANGES_REQUESTED** [review]
Looks good overall, just some cleanups and avoiding disk.

**Comment by avaer** [comment]
Should make sure this isn't consuming the stream: https://github.com/webaverse/ethereum-bot/pull/11/files#diff-c6a1938c899065d78caa9fdf781aeae99f6262dddb9a7c77fd8d3536877ccaf6R896

Can probably just use `http[s].request`.

File: constants.js [hunk 8]
```diff
@@ -1,5 +1,6 @@
 exports.ethereumHost = 'ethereum.exokit.org';
 exports.usersTableName = 'users';
+exports.twitterUsersTableName = 'users';
```
Comment (avaer) [inline constants.js]: Not used anymore.

**Comment by avaer** [comment]
I'm gonna clean this up so we can ship it.

File: discordbot.js [hunk 9]
```diff
@@ -2302,6 +2301,7 @@ Keys (DM bot)
                                                 if (status) {
                                                     const description = '';
                                                     const extName = path.extname(file.name).slice(1);
+                                                    extName = extName === "jpeg" ? "jpg" : extName
```
Comment (avaer) [inline discordbot.js]: It might be best to do this on the frontend.

**Review by avaer — APPROVED** [review]

**Pull request merged**